File: fp-in-ocaml/answers/state/candy.ml
type input = Coin | Turn
type machine = { locked : bool; candies : int; coins : int }

(** Exercise 6.11: [State.Monad]を用いて以下のルールを満たすキャンディ販売機の
    振る舞いをシミュレートする関数[simulate_machine]を実装せよ。
    [simulate_machine]は入力リストを受け取って販売機の最終的な キャンディの個数とコインの枚数のペアを返す。

    ルール:
    - 販売機がロックされている([locked = true])とき、ノブを回し([Turn])ても反応しない
    - 販売機がロックされている([locked = true])とき、コインを投入する([Coin])と ロックが外れてコインが1枚増える
    - 販売機がロックされていない([locked = false])とき、ノブを回す([Turn])と ロックが掛かってキャンディが1個減る
    - 販売機がロックされていない([locked = false])とき、コインを投入し([Coin])ても反応しない
    - キャンディが残っていない([candies = 0])とき、コインを投入([Coin])しても ノブを回([Turn])しても反応しない *)
let simulate_machine (inputs : input list) : (machine, int * int) Monad.t =
  let open Monad in
  let open Syntax in
  let* _ =
    inputs
    |> traverse @@ fun i ->
       modify @@ fun m ->
       match (i, m) with
       (* ルール上は最後に記載されているが、最初にマッチして何も起きないことを記述。 *)
       | _, { candies = 0; _ } -> m
       (* locked = true の場合を見る。
          必要ないパラメータは[_]でまとめて捨てることができる。
        *)
       | Turn, { locked = true; _ } -> m
       | Coin, { locked = true; _ } ->
           (* record のアップデートには [with] が使える。
              @see <https://ocaml.org/manual/5.4/coreexamples.html#s:tut-recvariants> Records and variants
            *)
           { m with locked = false; coins = m.coins + 1 }
       (* 以降は locked = false
          実装上は[m]のパターンを記述せずとも問題ないが、今回はルールをそのまま書いておく。
        *)
       | Turn, { locked = false; _ } ->
           { m with locked = true; candies = m.candies - 1 }
       | Coin, { locked = false; _ } -> m
  in
  let+ m = get in
  (m.coins, m.candies)

File: fp-in-clojure/src/fp_in_clojure/answers/state/candy.clj
(ns fp-in-clojure.answers.state.candy
  (:require
   [clojure.spec.alpha :as s]
   [fp-in-clojure.answers.state.candy.machine :as-alias machine]
   [fp-in-clojure.answers.state.state :as state]))

;; Exercise 6.11: Stateを用いて以下のルールを満たすキャンディ販売機の振る舞いをシミュレートする関数 `simulate-machine` を実装せよ。
;; `simulate-machine` は入力リストを受け取って販売機の最終的なコインの枚数とキャンディの個数のペアを返す。
;; ルール:
;;   - 販売機がロックされている(`locked? = true`)とき、ノブを回し(:turn)ても販売機は反応しない
;;   - 販売機がロックされている(`locked? = true`)とき、コインを投入する(:coin)と販売機のロックが外れてコインが1枚増える
;;   - 販売機がロックされていない(`locked? = false`)とき、ノブを回す(:turn)と販売機のロックが掛かってキャンディが1個減る
;;   - 販売機がロックされていない(`locked? = false`)とき、コインを投入し(:coin)ても販売機は反応しない
;;   - 販売機にキャンディが残っていない(`candies = 0`)とき、コインを投入し(:coin)てもノブを回し(:turn)ても販売機は反応しない

(def input #{:coin :turn})
(s/def ::machine/locked? boolean?)
(s/def ::machine/candies nat-int?)
(s/def ::machine/coins nat-int?)

(s/def ::machine/machine
  (s/keys :req [::machine/locked? ::machine/candies ::machine/coins]))

(s/fdef update-machine
  :args (s/cat :m ::machine/machine
               :i input)
  :ret ::machine/machine)

;; NOTE: 値による条件分岐を整理するためにマルチメソッドを利用している
;; ref. https://clojure.org/reference/multimethods
(defmulti ^:private update-machine
  (fn [machine input]
    (when-not (zero? (::machine/candies machine))
      (-> machine
          (select-keys [::machine/locked?])
          (assoc :input input)))))

(defmethod update-machine :default [m _] m)
(defmethod update-machine {:input :coin ::machine/locked? false} [m _] m)
(defmethod update-machine {:input :turn ::machine/locked? true} [m _] m)

(defmethod update-machine {:input :coin ::machine/locked? true} [m _]
  (-> m
      (assoc ::machine/locked? false)
      (update ::machine/coins inc)))

(defmethod update-machine {:input :turn ::machine/locked? false} [m _]
  (-> m
      (assoc ::machine/locked? true)
      (update ::machine/candies dec)))

;; Scala風の型表記: `simulate-machine: (inputs: Sequence[Input]) => State[Machine, (NatInt, NatInt)]`
(s/fdef simulate-machine
  :args (s/cat :inputs (s/coll-of input))
  :ret state/state?)

(defn simulate-machine [inputs]
  (->> inputs
       (state/traverse (comp state/modify (fn [input] #(update-machine % input))))
       (state/flat-map (fn [_]
                         (state/map (juxt ::machine/coins ::machine/candies)
                                    (state/get))))))

(comment
  (require '[clojure.spec.test.alpha :as stest])
  (stest/instrument)

  (state/run-state #::machine{:locked? true :candies 0 :coins 10} (simulate-machine [:coin]))
  (state/run-state #::machine{:locked? true :candies 0 :coins 10} (simulate-machine [:turn]))
  (state/run-state #::machine{:locked? false :candies 0 :coins 10} (simulate-machine [:coin]))
  (state/run-state #::machine{:locked? false :candies 0 :coins 10} (simulate-machine [:turn]))
  (state/run-state #::machine{:locked? true :candies 5 :coins 10} (simulate-machine [:coin]))
  (state/run-state #::machine{:locked? true :candies 5 :coins 10} (simulate-machine [:turn]))
  (state/run-state #::machine{:locked? false :candies 5 :coins 10} (simulate-machine [:coin]))
  (state/run-state #::machine{:locked? false :candies 5 :coins 10} (simulate-machine [:turn]))
  (state/run-state #::machine{:locked? true :candies 5 :coins 10} (simulate-machine [:coin :turn]))
  )
